src/base/extensible_array.hpp

```diff
@@ -0,0 +1,198 @@
+ *        ...
+ *        if (offset + size_of_data > capacity)           // size_of_data can vary from 0 to 16k, but will usually be
+ *                                                        // small enough
+ *          {
+ *            int new_capacity = capacity;
+ *            while (new_capacity < offset + size_of_data)    // double until we can cover
+ *              {
+ *                new_capacity = 2 * new_capacity
+ *              }
+ *
+ *            buffer_p = db_private_realloc (thread_p, buffer_p, new_capacity);
+ *          }
+ *        memcpy (buffer_p, page_ptr, size_of_data);
+ *        ...
+ *
+ *      end:
+ *        // delete if dynamically allocated
+ *        if (buffer_p != buffer)
+ *          {
+ *            db_private_free (thread_p, buffer_p);
+ *          }
+ *
+ *
+ * When not to use:
+ *
+ *    1. When memory requirements can be predicted and are reasonably small - use static array instead.
+ *
+ *    2. When performance is not an issue and dynamic allocation is affordable. Use std::string or std::vector instead.
+ *
+ *
+ * How to use:
+ *
+ *    1. Include extensible_array.cpp, where the template class implementation is found.
+ *
+ *    2. declare static variable with a default size:
+ *        const int BUF_SIZE = 64;
+ *        db_private_allocator<char> char_private_allocator (thread_p);
+ *        extensible_array<char, BUF_SIZE, db_private_allocator<char> > buffer(char_private_allocator);
+ *
+ *    2. append the required buffer.
+ *        buffer.append (REINTERPRET_CAST (char *, &header), sizeof (header));
+ *        buffer.append (page_ptr, size_of_data);  // size_of_data can vary from 0 to 16k, but will usually be
+ *                                                 // small enough
+ *
+ *    3. resources are freed automatically once variable scope ends.
+ *
+ *
+ * Templates:
+ *
+ *    1. T - buffer base type. in most cases a primitive is expected, but the buffer can accept any class.
+ *
+ *    2. Size - buffer static size. It is recommended to provide a value that covers most cases.
+ *
+ *    3. [Optional] Allocator - which allocator to be used for dynamic allocation. Default is standard allocator.
+ *
+ *
+ * Implementation details:
+ *
+ *    A std::array stores the static buffer and a std::vector the dynamic buffer. When static buffer size limit is
+ *    reached the vector is allocated. Size is always doubled to cover the new requirements. Vector is resized whenever
+ *    necessary.
+ *    There are two operations: append and copy (which starts over). They can accept an iterator (which is consumed
+ *    from begin to end), or start/end iterators.
+ *
+ *
+ * Notes:
+ *
+ *    If possible, provide a maximum buffer size to avoid uncontrolled allocations.
+ *
+ *
+ * TODO:
+ *
+ *    Add alignment. With this feature, we may even replace OR_BUF.
+ *
+ *    Should we handle out of memory?
+ */
+template <typename T, size_t Size, typename Allocator = std::allocator<T> >
```

Comment:
Thanks for the suggestion, I did move the memory management part from extensible_array to contiguous_memory_buffer. I could not use a generic allocator (even composed of static/dynamic allocators) since all I needed was one contiguous memory area.

Extensible array is now a wrapper that knows to append/copy data in one contiguous memory buffer.